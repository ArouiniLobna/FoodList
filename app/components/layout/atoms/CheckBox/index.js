import React, { memo, useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

function CheckBox({ className, checked, label, dataName, onChange }) {
  const [isChecked, seIsChecked] = useState(checked);

  const handleOnChange = e => {
    onChange(e.target.getAttribute('data-name'), !isChecked);
    seIsChecked(!isChecked);
  };
  return (
    <HiddenLabel>
      <CheckboxContainer className={className}>
        <HiddenCheckbox
          data-name={dataName}
          checked={isChecked}
          onChange={handleOnChange}
        />
        <StyledCheckbox checked={isChecked}>
          <Icon viewBox="0 0 24 24">
            <polyline points="20 6 9 17 4 12" />
          </Icon>
        </StyledCheckbox>
      </CheckboxContainer>
      <span>{label}</span>
    </HiddenLabel>
  );
}

CheckBox.propTypes = {
  className: PropTypes.string, // custom class name
  checked: PropTypes.bool.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  dataName: PropTypes.string, // the property name that needs to be updated in the json
};

const HiddenLabel = styled.label`
  width: 16px;
  height: 16px;
  line-height: 1;
  display: inline-block;
  & span {
    opacity: 0;
    position: absolute;
  }
`;
const CheckboxContainer = styled.div`
  display: inline-block;
  vertical-align: middle;
  width: 100%;
  height: 100%;
`;

const Icon = styled.svg`
  fill: none !important;
  stroke: white;
  stroke-width: 2px;
  vertical-align: unset;
`;
// Hide checkbox visually but remain accessible to screen readers.
const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  &:focus + div {
    box-shadow: 0 0 0 3px #bfc8d9;
  }
`;

const StyledCheckbox = styled.div`
  display: inline-block;
  width: 100%;
  height: 100%;
  background: ${props => (props.checked ? '#388e3c' : '#fff')};
  border-radius: 3px;
  transition: all 150ms;
  border: ${props =>
    props.checked ? ' 1px solid #388e3c' : ' 1px solid #7d90b2'};

  ${Icon} {
    visibility: ${props => (props.checked ? 'visible' : 'hidden')};
  }
`;

export default memo(CheckBox);
