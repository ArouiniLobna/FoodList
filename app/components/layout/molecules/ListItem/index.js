/**
 *
 * ListItem
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import CheckBox from 'components/layout/atoms/CheckBox';
import styled from 'styled-components';

function ListItem({ item, setCheckedStatus, itemKey }) {
  const handleSetCheckedStatus = (propertyName, isChecked) => {
    setCheckedStatus(itemKey, propertyName, isChecked);
  };
  return (
    <SListItem
      role="group" // as part of accessibilty we need to group the checkboxes
      aria-labelledby={`label-${itemKey}`}
      isHealthy={item.isHealthy}
      isDelicious={item.isDelicious}
    >
      <td>
        <h5 id={`label-${itemKey}`}>{item.label}</h5>
      </td>
      <td>
        <CheckBox
          checked={item.isDelicious}
          label="is Delicious"
          onChange={handleSetCheckedStatus}
          dataName="isDelicious"
        />
      </td>
      <td>
        <CheckBox
          checked={item.isHealthy}
          label="is Healthy"
          onChange={handleSetCheckedStatus}
          dataName="isHealthy"
        />
      </td>
    </SListItem>
  );
}

const SListItem = styled.tr`
  padding: 10px;
  display: flex;
  align-items: center;
  background-color: ${props => (props.isHealthy ? '#E0EEE0' : 'transparent')};
  transition: 1s;
  h5 {
    margin: 0;
    transition: 1s;
    text-indent: ${props => (props.isDelicious ? '30px' : '0')};
  }
`;

ListItem.propTypes = {
  setCheckedStatus: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired,
  itemKey: PropTypes.string.isRequired, // this the food item unique identifier
};

export default memo(ListItem);
