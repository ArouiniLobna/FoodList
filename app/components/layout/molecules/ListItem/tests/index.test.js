/**
 *
 * Tests for ListItem
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { mount } from 'enzyme';
import CheckBox from 'components/layout/atoms/CheckBox';

import ListItem from '../index';

describe('<ListItem />', () => {
  let container;
  let props;

  beforeEach(() => {
    props = {
      itemKey: '3434',
      item: { isDelicious: false, isHealthy: true, label: 'dummy food' },
      setCheckedStatus: jest.fn(),
    };
    container = mount(<ListItem {...props} />);
  });

  it('Expect to render 3 <td />" ', () => {
    expect(container.find('td').length).toBe(3);
  });
  it('Expect to render the item label within a h5" ', () => {
    expect(container.find('h5').text()).toEqual(props.item.label);
  });

  it('Expect to render 2 CheckBox" ', () => {
    expect(container.find(CheckBox).length).toBe(2);
  });
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<ListItem {...props} />);
    expect(firstChild).toMatchSnapshot();
  });
});
