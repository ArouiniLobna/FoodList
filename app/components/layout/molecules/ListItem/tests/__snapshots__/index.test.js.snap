// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> Should render and match the snapshot 1`] = `
<tr
  aria-labelledby="label-3434"
  class="ListItem__SListItem-y53y74-0 bYNREK"
  role="group"
>
  <td>
    <h5
      id="label-3434"
    >
      dummy food
    </h5>
  </td>
  <td>
    <label
      class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
    >
      <div
        class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
      >
        <input
          class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
          data-name="isDelicious"
          type="checkbox"
        />
        <div
          class="CheckBox__StyledCheckbox-sc-3ds2-4 eVKOMr"
        >
          <svg
            class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
            viewBox="0 0 24 24"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
        </div>
      </div>
      <span>
        is Delicious
      </span>
    </label>
  </td>
  <td>
    <label
      class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
    >
      <div
        class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
      >
        <input
          checked=""
          class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
          data-name="isHealthy"
          type="checkbox"
        />
        <div
          class="CheckBox__StyledCheckbox-sc-3ds2-4 gjKREU"
        >
          <svg
            class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
            viewBox="0 0 24 24"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
        </div>
      </div>
      <span>
        is Healthy
      </span>
    </label>
  </td>
</tr>
`;
